src/pyfmi/fmi2.pyx

```diff
@@ -0,0 +1,5256 @@
+
+# distutils: define_macros=NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION
+
+import os
+import logging
+cimport cython
+
+import numpy as np
+cimport numpy as np
+cimport pyfmi.fmil_import as FMIL
+cimport pyfmi.fmil2_import as FMIL2
+cimport pyfmi.fmi_base as FMI_BASE
+
+from collections import OrderedDict
+from pyfmi.fmi_base import (
+    PyEventInfo, enable_caching, FMI_DEFAULT_LOG_LEVEL, check_fmu_args
+)
+
+from pyfmi.exceptions import (
+    FMUException, InvalidBinaryException, InvalidXMLException, 
+    InvalidVersionException
+)
+from pyfmi.common.core import create_temp_dir
+
+from numpy cimport PyArray_DATA
+
+import scipy.sparse as sps
+
+int   = np.int32
+np.int = np.int32
+
+# Basic flags related to FMI
+FMI2_TRUE = FMIL2.fmi2_true
+FMI2_FALSE = FMIL2.fmi2_false
+
+# Status
+FMI2_DO_STEP_STATUS       = FMIL2.fmi2_do_step_status
+FMI2_PENDING_STATUS       = FMIL2.fmi2_pending_status
+FMI2_LAST_SUCCESSFUL_TIME = FMIL2.fmi2_last_successful_time
+FMI2_TERMINATED           = FMIL2.fmi2_terminated
+
+# Types
+FMI2_REAL        = FMIL2.fmi2_base_type_real
+FMI2_INTEGER     = FMIL2.fmi2_base_type_int
+FMI2_BOOLEAN     = FMIL2.fmi2_base_type_bool
+FMI2_STRING      = FMIL2.fmi2_base_type_str
+FMI2_ENUMERATION = FMIL2.fmi2_base_type_enum
+
+# Variability
+FMI2_CONSTANT   = FMIL2.fmi2_variability_enu_constant
+FMI2_FIXED      = FMIL2.fmi2_variability_enu_fixed
+FMI2_TUNABLE    = FMIL2.fmi2_variability_enu_tunable
+FMI2_DISCRETE   = FMIL2.fmi2_variability_enu_discrete
+FMI2_CONTINUOUS = FMIL2.fmi2_variability_enu_continuous
+FMI2_UNKNOWN    = FMIL2.fmi2_variability_enu_unknown
+
+# Causality
+FMI2_INPUT     = FMIL2.fmi2_causality_enu_input
+FMI2_OUTPUT    = FMIL2.fmi2_causality_enu_output
+FMI2_PARAMETER = FMIL2.fmi2_causality_enu_parameter
+FMI2_CALCULATED_PARAMETER = FMIL2.fmi2_causality_enu_calculated_parameter
+FMI2_LOCAL       = FMIL2.fmi2_causality_enu_local
+FMI2_INDEPENDENT = FMIL2.fmi2_causality_enu_independent
+
+# Dependency
+FMI2_KIND_DEPENDENT = FMIL2.fmi2_dependency_factor_kind_dependent
+FMI2_KIND_CONSTANT  = FMIL2.fmi2_dependency_factor_kind_constant
+FMI2_KIND_FIXED     = FMIL2.fmi2_dependency_factor_kind_fixed
+FMI2_KIND_TUNABLE   = FMIL2.fmi2_dependency_factor_kind_tunable
+FMI2_KIND_DISCRETE  = FMIL2.fmi2_dependency_factor_kind_discrete
+
+# INITIAL
+FMI2_INITIAL_EXACT      = 0
+FMI2_INITIAL_APPROX     = 1
+FMI2_INITIAL_CALCULATED = 2
+FMI2_INITIAL_UNKNOWN    = 3
+
+# Jacobian approximation
+DEF FORWARD_DIFFERENCE = 1
+DEF CENTRAL_DIFFERENCE = 2
+FORWARD_DIFFERENCE_EPS = (np.finfo(float).eps)**0.5
+CENTRAL_DIFFERENCE_EPS = (np.finfo(float).eps)**(1/3.0)
+
+# Flags for evaluation of FMI Jacobians
+# Evaluate Jacobian w.r.t. states.
+FMI_STATES = 1
+# Evaluate Jacobian w.r.t. inputs.
+FMI_INPUTS = 2
+# Evaluate Jacobian of derivatives.
+FMI_DERIVATIVES = 1
+# Evaluate Jacobian of outputs.
+FMI_OUTPUTS = 2
+
+# TODO: Copied from utils temporarily; to avoid circular dependencies
```

Comment:
fmi_util.pyx has some unpleasant circular dependencies with the fmi*.pyx files, this is the only place this function is used though.

Copied it here for now to get rid of the circular dependency.